lib/Sema/TypeCheckEffects.cpp

```diff
@@ -3570,16 +3570,40 @@ class CheckEffectsCoverage : public EffectsHandlingWalker<CheckEffectsCoverage>
       if (const AnyTryExpr *tryExpr = dyn_cast<AnyTryExpr>(autoClosure->getSingleExpressionBody()))
         awaitInsertLoc = tryExpr->getSubExpr()->getStartLoc();
     }
+    
+    auto findShorthandOptionalBinding = [&](Expr *anchor) -> VarDecl * {
+      VarDecl *varDecl = nullptr;
+      if (!anchor->getType()->isOptional())
+        return nullptr;
+      anchor->forEachChildExpr([&](Expr *expr) -> Expr * {
+        if (auto *declRef = dyn_cast<DeclRefExpr>(expr)) {
+          if (declRef && declRef->isImplicit()) {
```

Comment:
Did all the tests passed? I am just curious to see if this can indeed detect narrowly the case we want for diagnostic and will not return unwanted cases such as `let a = anAsyncFuncThatReturnsOptional(value)` where the `CallExpr` returns optional and value is a optional variable. Although maybe `declRef->isImplicit()` may cover that, I wonder if we can be more specific on the context where we look for a `OptionalEvaluationExpr` for example instead of `anchor->getType()->isOptional()`. But that is just a guess for my side, maybe worth looking into it.

Reply:
The tests do pass, I added another test case based off of `let a = anAsyncFuncThatReturnsOptional(value)` it passes as well. I see what you're saying but I'm not sure if `OptionalEvaluationExpr` applies here. For example, running `swift -dump-ast -O` on the following:
```swift
func test() async {
  async let result: String? = nil
  if let result {}
}
```
we get (not including all output here btw): 
```
(if_stmt range=[tmp.swift:3:3 - line:3:18]
        (pattern
          (pattern_optional_some implicit type='String?'
            (pattern_let implicit type='String'
              (pattern_named type='String' 'result')))
          (declref_expr implicit type='String?' location=tmp.swift:3:10 range=[tmp.swift:3:10 - line:3:10] decl=tmp.(file).test().result@tmp.swift:2:13 function_ref=unapplied))
        (brace_stmt range=[tmp.swift:3:17 - line:3:18])))))
```

You would get `OptionalEvaluationExpr` if you wrote something like:
```swift
struct S { let x: String }
func foo() {
  let s: S? = nil
  // the optional chaining here
  if let x = s?.x {
    print(x)
  }
}
```

Reply:
Yeah, it doesn't matter whether is an optiona_evaluation_expr or pattern_optional_some, I just remember wrong AST node, but point is we can be more explicit in the check to avoid false positives, checking only for "type is optional" would not guarantee we are in a let optional binding, so narrowing down to `pattern_optional_some` seems to be safer to the kind of pattern we want to detect here.

Reply:
I was looking into this today, seems like we should be able to check for the pattern, but from this part of the code I haven't figured it out yet. Was trying `varDecl->getParentPattern()` but that doesn't seem to be what we want. Will continue trying of course, just wanted to give an update.